File: 백준/Gold/1339. 단어 수학/단어 수학.py
import heapq
import sys
from collections import defaultdict, deque


# 단어 수학.
# n개의 단어, 각 단어 알파벳 대문자.
# 각 문자를 0~9 중 하나로 바꿔서 N개의 수를 합하는 문제.같은 알파벳은 같은 숫자로 바꿔야 함,같은 숫자 겹치면 안됨
# N개의 단어 -> 그 수의 합 최대

n = int(input())
data = []
map = defaultdict(int)
for _ in range(n):
    words = input() # 단어 입력
    data.append(words)
    for i in range(len(words)):
        now = words[i]
        map[now] += 10 ** (len(words) - i - 1)

# 값이 큰 순서대로 큰 값 부여 : value 기준
map = dict(sorted(map.items(), key = lambda x : x[1], reverse = True)) # value를 기준으로 내림차순 정렬
num_dict = defaultdict(int)
num = 9
for key, val in map.items():
    num_dict[key] = num
    num -= 1

# print(num_dict.items())
# print(data)
res = 0
for words in data:
    now = ""
    for word in words:
        now += str(num_dict[word])
    # print(now)
    res += int(now)
print(res)

File: 백준/Gold/1339. 단어 수학/단어 수학.java
import java.io.*;
import java.util.*;



public class Main {

    static int n;
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static List<String> data = new ArrayList<>();
    static Map<Character, Integer> map = new HashMap();
    static Map<Character, Integer> num_map = new HashMap<>();

    static String[] words;
    public static void main(String[] args) throws IOException {

        n = Integer.parseInt(br.readLine());
        words = new String[n];
        for (int i = 0; i < n; i++) {
            String data = br.readLine();
            words[i] = data;
            for (int j = 0; j < data.length(); j++) {
                char c = data.charAt(j); // ! 변환
//                System.out.println(Math.pow(10, data.length() - j - 1));
                map.put(c, map.getOrDefault(c, 0) + (int) Math.pow(10, data.length() - j - 1));
            }
        }
//        System.out.println("끝나고확인 : " + map);
        // ! 알파벳의 가치를 기준으로 내림차순 정렬
        // ! List로 정렬 후 사용
//        List<Character> chars = new ArrayList<>(map.keySet());
//        Collections.sort(chars, (a, b) -> map.get(b) - map.get(a));
        // ! Collections.sort를 통해 List형태로 Map 가져오기
        List<Map.Entry<Character, Integer>> entryList = new ArrayList<>(map.entrySet());
        // ! 이 entrySet을 정렬하는 방법 사용
        // ! entry의 값을 기준으로 내림차순 정렬 진행.
        Collections.sort(entryList, (a, b) -> b.getValue() - a.getValue());
        // ! 알파벳에 숫자 할당
        int num = 9;
//        for (char c : chars) {
//            num_map.put(c, num);
//            num -= 1;
//        }
        for (Map.Entry<Character, Integer> entry : entryList) {
            num_map.put(entry.getKey(), num);
            num -= 1;
        }
        // ! 정렬된 entryList를 순회하면서 각 엔트리의 키를 사용하여 num_map에 알파벳 숫자를 매핑
//        System.out.println(num_map);
        // ! 최종 결과
        int res = 0;
        for (String word : words) {
            int value = 0;
            for (char c : word.toCharArray()) {
                value = value * 10 + num_map.get(c);
            }
            res += value;
        }
//        System.out.println(map);
//        System.out.println(num_map);
        System.out.println(res);
    }
}